dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/ReportRatesHelper.java

```diff
@@ -0,0 +1,129 @@
+import org.hisp.dhis.period.Period;
+import org.hisp.dhis.period.PeriodType;
+
+/**
+ * Just a helper class responsible for providing methods that hides specific
+ * Report Rates logic and calculation.
+ *
+ * @author maikel arabori
+ */
+public class ReportRatesHelper
+{
+    private ReportRatesHelper()
+    {
+    }
+
+    /**
+     * Use number of days for daily data sets as target, as query periods might
+     * often span/contain different numbers of days.
+     *
+     * @param periodIndex the index of the period in the "dataRow".
+     * @param timeUnits the time unit size found in the current DataQueryParams.
+     *        See {@link #DataHandler.getTimeUnits(DataQueryParams)}.
+     * @param dataRow the current dataRow, based on the key map built by
+     *        {@link #DataHandler.getAggregatedCompletenessTargetMap(DataQueryParams)).
+     * @param target the current value of the respective key ("dataRow"). See
+     *        {@link #DataHandler.getAggregatedCompletenessTargetMap(DataQueryParams).
+     * @param queryPt the filter period in the current "dataRow". See
+     *        {@link PeriodType#getPeriodTypeFromIsoString}.
+     * @param dataSetPt the dataset period.
+     * @param filterPeriods the filter "pe" in the params.
+     *
+     * @return the calculate target
+     */
+    public static Double getCalculatedTarget( final int periodIndex, final int timeUnits, final List<String> dataRow,
+        final Double target, final PeriodType queryPt, final PeriodType dataSetPt,
+        final List<DimensionalItemObject> filterPeriods )
+    {
+        if ( dataSetPt.equalsName( NAME ) )
+        {
+            boolean hasPeriodInDimension = periodIndex != -1;
+
+            // If we enter here, it means there is a "pe" in the dimension
+            // parameter.
+            if ( hasPeriodInDimension )
+            {
+                return getTargetFromDimension( periodIndex, timeUnits, dataRow, target );
+            }
+            else
+            {
+                // If we reach here, it means that we should have a "pe"
+                // dimension as filter parameter.
+                return getTargetFromFilter( timeUnits, target, filterPeriods );
+            }
+        }
+        else
+        {
+            return consolidateTarget( timeUnits, target, queryPt.getPeriodSpan( dataSetPt ) );
+        }
+    }
+
+    private static Double getTargetFromDimension( final int periodIndex, final int timeUnits,
+        final List<String> dataRow, final Double target )
+    {
+        final Period period = PeriodType.getPeriodFromIsoString( dataRow.get( periodIndex ) );
+
+        return consolidateTarget( timeUnits, target, period.getDaysInPeriod() );
+    }
+
+    private static Double getTargetFromFilter( final int timeUnits, final Double target,
+        final List<DimensionalItemObject> filterPeriods )
+    {
+        final List<DimensionalItemObject> periods = filterPeriods;
```

Comment:
What is the reason for periods variable?

Reply:
Good catch, not needed. I will remove it soon.

Reply:
Done